# Data Consistency Checks and Descriptive Analysis

## This script contains the following points:

### 01. Importing Libraries and Loading CSV Files
### 02. Cleaning df_results
### 03. Cleaning df_goalscorers
### 04. Cleaning df_shootouts
### 05. Analyzing df_results
### 06. Exporting Cleaned Datasets

In [1]:
# Importing Python libraries
import pandas as pd
import numpy as np
import os

# Creating path variable
path = r'C:\Users\widne\Documents\CareerFoundry Exercises\Data_Immersion\Achievement 6\06-2024 International Football Matches'

# Loading csv files
df_results = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'results.csv'), index_col=False)
df_goalscorers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'goalscorers.csv'), index_col=False)
df_shootouts = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'shootouts.csv'), index_col=False)

In [3]:
df_results.shape

(47126, 9)

In [2]:
# Checking df_results
df_results.head()

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  Scotland   England         0.0         0.0   Friendly  Glasgow   
1  1873-03-08   England  Scotland         4.0         2.0   Friendly   London   
2  1874-03-07  Scotland   England         2.0         1.0   Friendly  Glasgow   
3  1875-03-06   England  Scotland         2.0         2.0   Friendly   London   
4  1876-03-04  Scotland   England         3.0         0.0   Friendly  Glasgow   

    country  neutral  
0  Scotland    False  
1   England    False  
2  Scotland    False  
3   England    False  
4  Scotland    False

In [3]:
# Checking df_goalscorers
df_goalscorers.head()

date  home_team away_team       team            scorer  minute  \
0  1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    44.0   
1  1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    55.0   
2  1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    70.0   
3  1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    75.0   
4  1916-07-06  Argentina     Chile  Argentina     Alberto Ohaco     2.0   

  own_goal penalty  
0    False   False  
1    False   False  
2    False   False  
3    False   False  
4    False   False

In [4]:
# Checking df_shootouts
df_shootouts.head()

date    home_team         away_team       winner first_shooter
0  1967-08-22        India            Taiwan       Taiwan           NaN
1  1971-11-14  South Korea  Vietnam Republic  South Korea           NaN
2  1972-05-07  South Korea              Iraq         Iraq           NaN
3  1972-05-17     Thailand       South Korea  South Korea           NaN
4  1972-05-19     Thailand          Cambodia     Thailand           NaN

### 02. Cleaning df_results

#### Mixed-Type Data

In [5]:
# Checking for mixed-type data
for col in df_results.columns.tolist():
    weird = (df_results[[col]].map(type) != df_results[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_results[weird]) > 0:
        print(col)

home_team
away_team


The home_team and away_team columns have mixed data types. For cleaning I will set these columns to string values

In [6]:
# Setting home_team column to string type
df_results['home_team'] = df_results['home_team'].astype('str')

In [7]:
# Setting away_team column to string type
df_results['away_team'] = df_results['away_team'].astype('str')

#### Missing Values

In [8]:
df_results.isnull().sum()

date           0
home_team      0
away_team      0
home_score    51
away_score    51
tournament     0
city           0
country        0
neutral        0
dtype: int64

In [9]:
df_results_nan = df_results[df_results['home_score'].isnull() == True]

In [10]:
df_results_nan

date       home_team       away_team  home_score  away_score  \
47075  2024-06-14         Germany        Scotland         NaN         NaN   
47076  2024-06-15         Hungary     Switzerland         NaN         NaN   
47077  2024-06-15           Spain         Croatia         NaN         NaN   
47078  2024-06-15           Italy         Albania         NaN         NaN   
47079  2024-06-16        Slovenia         Denmark         NaN         NaN   
47080  2024-06-16          Serbia         England         NaN         NaN   
47081  2024-06-16          Poland     Netherlands         NaN         NaN   
47082  2024-06-17         Austria          France         NaN         NaN   
47083  2024-06-17         Romania         Ukraine         NaN         NaN   
47084  2024-06-17         Belgium        Slovakia         NaN         NaN   
47085  2024-06-18          Turkey         Georgia         NaN         NaN   
47086  2024-06-18        Portugal  Czech Republic         NaN         NaN   
47087  2024-06-19         Germany         Hungary         NaN         NaN   
47088  2024-06-19        Scotland     Switzerland         NaN         NaN   
47089  2024-06-19         Croatia         Albania         NaN         NaN   
47090  2024-06-20           Spain           Italy         NaN         NaN   
47091  2024-06-20        Slovenia          Serbia         NaN         NaN   
47092  2024-06-20         Denmark         England         NaN         NaN   
47093  2024-06-21          Poland         Austria         NaN         NaN   
47094  2024-06-21     Netherlands          France         NaN         NaN   
47095  2024-06-21        Slovakia         Ukraine         NaN         NaN   
47096  2024-06-22         Belgium         Romania         NaN         NaN   
47097  2024-06-22         Georgia  Czech Republic         NaN         NaN   
47098  2024-06-22          Turkey        Portugal         NaN         NaN   
47099  2024-06-23         Germany     Switzerland         NaN         NaN   
47100  2024-06-23        Scotland         Hungary         NaN         NaN   
47101  2024-06-24         Albania           Spain         NaN         NaN   
47102  2024-06-24         Croatia           Italy         NaN         NaN   
47103  2024-06-25         England        Slovenia         NaN         NaN   
47104  2024-06-25         Denmark          Serbia         NaN         NaN   
47105  2024-06-25     Netherlands         Austria         NaN         NaN   
47106  2024-06-25          France          Poland         NaN         NaN   
47107  2024-06-26        Slovakia         Romania         NaN         NaN   
47108  2024-06-26         Ukraine         Belgium         NaN         NaN   
47109  2024-06-26         Georgia        Portugal         NaN         NaN   
47110  2024-06-26  Czech Republic          Turkey         NaN         NaN   
47111  2024-06-29             nan             nan         NaN         NaN   
47112  2024-06-29             nan             nan         NaN         NaN   
47113  2024-06-30             nan             nan         NaN         NaN   
47114  2024-06-30             nan             nan         NaN         NaN   
47115  2024-07-01             nan             nan         NaN         NaN   
47116  2024-07-01             nan             nan         NaN         NaN   
47117  2024-07-02             nan             nan         NaN         NaN   
47118  2024-07-02             nan             nan         NaN         NaN   
47119  2024-07-05             nan             nan         NaN         NaN   
47120  2024-07-05             nan             nan         NaN         NaN   
47121  2024-07-06             nan             nan         NaN         NaN   
47122  2024-07-06             nan             nan         NaN         NaN   
47123  2024-07-09             nan             nan         NaN         NaN   
47124  2024-07-10             nan             nan         NaN         NaN   
47125  2024-07-14             nan             nan         NaN         NaN   

      to

All null values are for matches that have not been played at. I intend to only analyze historical data so I am dropping the null values from the data.

In [11]:
# Dropping null values from data
df_results_clean = df_results[df_results['home_score'].isnull()==False]

In [12]:
# Checking if all nulls are dropped
df_results_clean.isnull().sum()

date          0
home_team     0
away_team     0
home_score    0
away_score    0
tournament    0
city          0
country       0
neutral       0
dtype: int64

In [13]:
# Comparing shape of df_results and df_results_clean
df_results.shape

(47126, 9)

In [14]:
df_results_clean.shape

(47075, 9)

51 records were dropped as expected.

#### Finding Duplicates

In [15]:
# Checking df_results for duplicates
df_results_dups = df_results_clean[df_results_clean.duplicated()]

In [16]:
df_results_dups

Empty DataFrame
Columns: [date, home_team, away_team, home_score, away_score, tournament, city, country, neutral]
Index: []

No duplicates were found.

### 03. Cleaning df_goalscorers 
#### Mixed-Type Data

In [17]:
# Checking df_goalscorers for mixed data types
for col in df_goalscorers.columns.tolist():
    weird = (df_goalscorers[[col]].map(type) != df_goalscorers[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_goalscorers[weird]) > 0:
        print(col)

scorer
own_goal
penalty


In [18]:
df_goalscorers.isnull().sum()

date           0
home_team      0
away_team      0
team           0
scorer        50
minute       263
own_goal       2
penalty        2
dtype: int64

In [19]:
df_nan_scorers = df_goalscorers[df_goalscorers['scorer'].isnull()==True]

In [20]:
df_nan_scorers

date        home_team         away_team              team scorer  \
8681  1980-02-24          Vanuatu  Papua New Guinea  Papua New Guinea    NaN   
8682  1980-02-24          Vanuatu  Papua New Guinea  Papua New Guinea    NaN   
8683  1980-02-24          Vanuatu  Papua New Guinea  Papua New Guinea    NaN   
8684  1980-02-24          Vanuatu  Papua New Guinea  Papua New Guinea    NaN   
8685  1980-02-24          Vanuatu  Papua New Guinea           Vanuatu    NaN   
8686  1980-02-24          Vanuatu  Papua New Guinea           Vanuatu    NaN   
8687  1980-02-24          Vanuatu  Papua New Guinea           Vanuatu    NaN   
8688  1980-02-25             Fiji   Solomon Islands              Fiji    NaN   
8689  1980-02-25             Fiji   Solomon Islands              Fiji    NaN   
8690  1980-02-25             Fiji   Solomon Islands              Fiji    NaN   
8691  1980-02-25             Fiji   Solomon Islands   Solomon Islands    NaN   
8709  1980-02-26    New Caledonia           Vanuatu     New Caledonia    NaN   
8710  1980-02-26    New Caledonia           Vanuatu     New Caledonia    NaN   
8711  1980-02-26    New Caledonia           Vanuatu     New Caledonia    NaN   
8712  1980-02-26    New Caledonia           Vanuatu     New Caledonia    NaN   
8713  1980-02-26    New Caledonia           Vanuatu           Vanuatu    NaN   
8714  1980-02-26    New Caledonia           Vanuatu           Vanuatu    NaN   
8715  1980-02-26    New Caledonia           Vanuatu           Vanuatu    NaN   
8728  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8729  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8730  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8731  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8732  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8733  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8734  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8735  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8736  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8737  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8738  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8739  1980-02-27  Solomon Islands            Tahiti            Tahiti    NaN   
8740  1980-02-27  Solomon Islands            Tahiti   Solomon Islands    NaN   
8742  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8743  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8744  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8745  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8746  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8747  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8748  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8749  1980-02-28    New Caledonia  Papua New Guinea     New Caledonia    NaN   
8750  1980-02-29             Fiji            Tahiti            Tahiti    NaN   
8751  1980-02-29             Fiji            Tahiti            Tahiti    NaN   
8752  1980-02-29             Fiji            Tahiti            Tahiti    NaN   
8753  1980-02-29             Fiji            Tahiti            Tahiti    NaN   
8754  1980-02-29             Fiji            Tahiti            Tahiti    NaN   
8755  1980-02-29             Fiji            Tahiti            Tahiti    NaN   
8756  1980-02-29             Fiji            Tahiti              Fiji    NaN   
8757  1980-02-29             Fiji            Tahiti              Fiji    NaN   
8758  1980-02-29             Fiji            Tahiti              Fiji    NaN   
8962  1980-09-23         Malaysia             Qatar             Qatar    NaN   
8964  1980-0

NaN values in the scorers column seem to be causing the mixed data type. To address this I will make the null values in the scorer column to have an 'Unknown' string value.

#### Missing Values

In [21]:
# Changing nulls in scorer column to have 'Unknown' string value
df_goalscorers.scorer = df_goalscorers.scorer.fillna('Unknown')

In [22]:
# Checking for no null values in scorer column
df_goalscorers.isnull().sum()

date           0
home_team      0
away_team      0
team           0
scorer         0
minute       263
own_goal       2
penalty        2
dtype: int64

In [23]:
df_nan_minute = df_goalscorers[df_goalscorers['minute'].isnull()==True]

In [24]:
df_nan_minute

date     home_team         away_team          team  \
3331   1960-10-16        Taiwan  Vietnam Republic        Taiwan   
4043   1963-11-26         Ghana          Ethiopia         Ghana   
4044   1963-11-26         Ghana          Ethiopia         Ghana   
4047   1963-11-28      Ethiopia           Tunisia      Ethiopia   
4048   1963-11-28      Ethiopia           Tunisia      Ethiopia   
...           ...           ...               ...           ...   
16798  1997-03-27  Saudi Arabia        Bangladesh  Saudi Arabia   
16824  1997-03-29        Taiwan        Bangladesh        Taiwan   
16825  1997-03-29        Taiwan        Bangladesh    Bangladesh   
16826  1997-03-29        Taiwan        Bangladesh    Bangladesh   
16835  1997-03-31    Bangladesh          Malaysia      Malaysia   

                     scorer  minute own_goal penalty  
3331          Yiu Cheuk Yin     NaN    False   False  
4043          Edward Acquah     NaN    False   False  
4044          Edward Acquah     NaN    False   False  
4047         Mengistu Worku     NaN    False   False  
4048         Mengistu Worku     NaN    False   False  
...                     ...     ...      ...     ...  
16798    Abdullah Al-Dosari     NaN    False   False  
16824           Hsu Te Ming     NaN    False   False  
16825           Alfaz Ahmed     NaN    False   False  
16826    Imtiaz Ahmed Nakib     NaN    False   False  
16835  Zainal Abidin Hassan     NaN    False   False  

[263 rows x 8 columns]

There are 263 records with an unknown minute of the goal scored. Since this is only about 0.5% of the entire data, I will impute the mean of the minute scored to these values.

In [25]:
# Imputing mean of muinute to missing values in column
df_goalscorers.describe()

minute
count  43847.000000
mean      50.015212
std       26.354534
min        1.000000
25%       28.000000
50%       51.000000
75%       73.000000
max      122.000000

With a mean of 50 the null values will be imputed with that value.

In [26]:
df_goalscorers['minute'].fillna(50, inplace=True)

In [27]:
# Chekcing for missing values for minute column
df_goalscorers.isnull().sum()

date         0
home_team    0
away_team    0
team         0
scorer       0
minute       0
own_goal     2
penalty      2
dtype: int64

In [28]:
# Checking null values in own_goal column
df_nan_own_goal = df_goalscorers[df_goalscorers['own_goal'].isnull()==True]

In [29]:
df_nan_own_goal

date home_team away_team      team          scorer  minute  \
8961  1980-09-23  Malaysia     Qatar  Malaysia  Tukamin Bahari    50.0   
8962  1980-09-23  Malaysia     Qatar     Qatar         Unknown    50.0   

     own_goal penalty  
8961      NaN     NaN  
8962      NaN     NaN

Only 2 values are null in own_goal and penalty columns. I will impute the mode of each column to these missing values since these columns have boolean values.

In [30]:
df_goalscorers['own_goal'].value_counts(dropna=False)

own_goal
False    43302
True       806
NaN          2
Name: count, dtype: int64

Since False is the mode value of own_goal, the missing values will have False for own_goal

In [31]:
df_goalscorers['penalty'].value_counts(dropna=False)

penalty
False    41160
True      2948
NaN          2
Name: count, dtype: int64

Since False is the mode value of pentalty, the missing values will have False for penalty

In [32]:
# Replacing missing values in own_goal and pentalty to False
df_goalscorers['own_goal'].fillna(False, inplace=True)
df_goalscorers['penalty'].fillna(False, inplace=True)

In [33]:
# Checking that all missing values are addressed
df_goalscorers.isnull().sum()

date         0
home_team    0
away_team    0
team         0
scorer       0
minute       0
own_goal     0
penalty      0
dtype: int64

All missing values and mixed data types are addressed.

#### Finding Duplicates

In [34]:
# Checking df_goalscorers for duplicates
df_goalscorers_dups = df_goalscorers[df_goalscorers.duplicated()]

In [35]:
df_goalscorers_dups

date    home_team            away_team         team  \
4044   1963-11-26        Ghana             Ethiopia        Ghana   
4048   1963-11-28     Ethiopia              Tunisia     Ethiopia   
4052   1963-11-28      Nigeria                Sudan        Sudan   
4214   1965-03-14     Suriname  Trinidad and Tobago     Suriname   
5090   1968-01-16        Congo                Ghana        Ghana   
...           ...          ...                  ...          ...   
43273  2023-06-18  Netherlands                Italy        Italy   
43275  2023-06-18  Netherlands                Italy        Italy   
43277  2023-06-18  Netherlands                Italy  Netherlands   
43279  2023-06-18  Netherlands                Italy        Italy   
43281  2023-06-18  Netherlands                Italy  Netherlands   

                    scorer  minute  own_goal  penalty  
4044         Edward Acquah    50.0     False    False  
4048        Mengistu Worku    50.0     False    False  
4052      Nasr Eddin Abbas    50.0     False    False  
4214       Edmund Waterval    50.0     False    False  
5090             Osei Kofi    50.0     False    False  
...                    ...     ...       ...      ...  
43273     Federico Dimarco     6.0     False    False  
43275      Davide Frattesi    20.0     False    False  
43277      Steven Bergwijn    68.0     False    False  
43279      Federico Chiesa    72.0     False    False  
43281  Georginio Wijnaldum    89.0     False    False  

[121 rows x 8 columns]

There are 121 duplicate records

In [36]:
# Dropping duplicate values
df_goalscorers_clean = df_goalscorers.drop_duplicates()

In [37]:
# Checking that duplicates are dropped
df_goalscorers.shape

(44110, 8)

In [38]:
df_goalscorers_clean.shape

(43989, 8)

121 values have been dropped which is the exact amount of duplicates found.

### 04. Cleaning df_shootout
#### Mixed Type Data

In [39]:
# Checking df_shootout for mixed data types
for col in df_shootouts.columns.tolist():
    weird = (df_shootouts[[col]].map(type) != df_shootouts[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_shootouts[weird]) > 0:
        print(col)

first_shooter


In [40]:
df_shootouts.isnull().sum()

date               0
home_team          0
away_team          0
winner             0
first_shooter    414
dtype: int64

In [41]:
df_nan_takers = df_shootouts[df_shootouts['first_shooter'].isnull()==True]

In [42]:
df_nan_takers

date        home_team         away_team         winner  \
0    1967-08-22            India            Taiwan         Taiwan   
1    1971-11-14      South Korea  Vietnam Republic    South Korea   
2    1972-05-07      South Korea              Iraq           Iraq   
3    1972-05-17         Thailand       South Korea    South Korea   
4    1972-05-19         Thailand          Cambodia       Thailand   
..          ...              ...               ...            ...   
595  2022-09-25         Malaysia        Tajikistan     Tajikistan   
612  2023-07-09           Orkney         Greenland      Greenland   
615  2023-07-14           Malawi           Lesotho        Lesotho   
616  2023-07-16     South Africa            Malawi   South Africa   
621  2023-12-02  Solomon Islands     New Caledonia  New Caledonia   

    first_shooter  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
..            ...  
595           NaN  
612           NaN  
615           NaN  
616           NaN  
621           NaN  

[414 rows x 5 columns]

The null values in the first_shooter column are causing the mixed data type. To address this I will change the null values of the first_shooter column to 'Unknown' string value,

#### Missing Values

In [43]:
# Replacing null values in df_shootouts
df_shootouts.first_shooter = df_shootouts.first_shooter.fillna('Unknown')

In [44]:
# Checking if missing values have been resolved
df_shootouts.isnull().sum()

date             0
home_team        0
away_team        0
winner           0
first_shooter    0
dtype: int64

#### Finding Duplicates

In [45]:
# Checking df_shootouts for duplicates
df_shootouts_dups = df_shootouts[df_shootouts.duplicated()]

In [46]:
df_shootouts_dups

Empty DataFrame
Columns: [date, home_team, away_team, winner, first_shooter]
Index: []

No duplicate records were found

### 06. Exporting Cleaned Datasets 

In [ ]:
# Exporting cleaned datasets
df_results_clean.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'results_cleaned.pkl'))
df_goalscorers_clean.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'goalscorers_cleaned.pkl'))
df_shootouts.to_pickle